### PR TITLE
[feature#14654] alert-spi support prometheus alertmanager

### LINKED ISSUE #14654: [Feature][Alert] Add prometheus's Alertmanager
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

Prometheus is the most popular cloud-native monitoring and alerting system.
We hope to support sending alerts directly to prometheus's [Alertmanager](https://prometheus.io/docs/alerting/latest/alertmanager/).

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by xjlgod**
## Purpose of the pull request

To add alert-api for prometheus alertmanager close #14654.
The implementation principle is to request its alert interface through http client to generate an alert.
[api document](https://github.com/prometheus/alertmanager/blob/main/api/v2/openapi.yaml)

## Brief change log

a new moudle dolphinscheduler-alert-prometheus be added

## Verify this pull request

create alert instance to test it.
9cc27e4905f8721471eb14f138a8d39
3e577f7f5471e1bc3f567b5fb2752c5

### CODE REVIEW

**Review by songjianet — DISMISSED** [review]

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        if (generatorURL != null && !generatorURL.equals("")) {
+            alert.put("generatorURL", generatorURL);
+        }
+        List<Map<String, Object>> body = new ArrayList<>();
+        body.add(alert);
+        return JSONUtils.toJsonString(body);
+    }
+
+    private static CloseableHttpClient getDefaultClient() {
+        return HttpClients.createDefault();
```
Comment (Gallardot) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: https://github.com/apache/dolphinscheduler/blob/a5284e40242eea7597e7be27d62e3133546e7073/dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-http/src/main/java/org/apache/dolphinscheduler/plugin/alert/http/HttpSender.java#L110-L116

Based on the implementation of other plugins, it is recommended to use an HTTP client that supports retrying

Comment (xjlgod) [reply dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: Ok.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/test/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSenderTest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class AlertManagerSenderTest {
+
+    private static Map<String, String> config = new HashMap<>();
+
+    @BeforeEach
+    public void initConfig() {
+        config.put(AlertManagerConstants.NAME_ALERT_MANAGER_URL, "http://127.0.0.1:9093");
```
Comment (Gallardot) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/test/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSenderTest.java]: How does this unit test mock an AlertManager-Server and pass the test?

**Review by Gallardot — COMMENTED** [review]
Rest LGTM.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/test/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSenderTest.java (hunk 2 above)
Comment (xjlgod) [reply dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/test/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSenderTest.java]: I set the unit test result to false. If need, I can add test to mock an AlertManager-Server. I did not add it for there is no mock for third server in any other alert plugins.

Comment (Gallardot) [reply dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/test/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSenderTest.java]: Got it. LGTM.

**Review by Gallardot — APPROVED** [review]

**Comment by Gallardot** [comment]
@SbloodyS @Radeity @EricGao888 @qingwli  PTAL.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-api/src/main/java/org/apache/dolphinscheduler/alert/api/AlertInputTips.java [hunk 3]
```diff
@@ -50,7 +50,8 @@ public enum AlertInputTips {
     SECRET("please input secret", "请输入secret"),
     WECHAT_MENTION_USERS("use `|` to separate userIds and `@all` to everyone", "使用`|`来分割userId或使用`@all`来提到所有人"),
     WECHAT_AGENT_ID("please input agent id or chat id", "请输入agent id或chat id"),
-    ;
+    ANNOTATION("please input annotation in json form", "请输入json格式的annotation"),
+    GENERATOR_URL("please input GeneratorURL", "请输入生成地址");
```
Comment (qingwli) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-api/src/main/java/org/apache/dolphinscheduler/alert/api/AlertInputTips.java]:
```suggestion
    GENERATOR_URL("please input Generator URL", "请输入生成地址");
```

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        annotations = config.get(AlertManagerConstants.NAME_ALERT_MANAGER_ANNOTATIONS);
+    }
+
+    public AlertResult sendMessage(AlertData alertData) {
+        AlertResult alertResult;
+        try {
+            String resp = sendMsg(alertData);
+            return checkSendAlertManageMsgResult(resp);
+        } catch (Exception e) {
+            log.info("send prometheus alert manager alert msg  exception: {}", e.getMessage());
```
Comment (Radeity) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: Use `log.error` is better

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        AlertResult alertResult = new AlertResult();
+        alertResult.setStatus("false");
+
+        if (Objects.equals(resp, AlertManagerConstants.ALERT_SUCCESS)) {
+            alertResult.setStatus("true");
+            alertResult.setMessage("prometheus alert manager send success");
+            return alertResult;
+        }
+
+        alertResult.setMessage("prometheus alert manager send fail");
```
Comment (qingwli) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: Please set the code and response in `alertResult.setMessage`. It's easier to find the error reason.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            alertResult.setMessage("send prometheus alert manager alert fail.");
+        }
+        return alertResult;
+    }
+
+    private String sendMsg(AlertData alertData) throws IOException {
+        String v2Path = String.format("%s%s", this.url, AlertManagerConstants.ALERT_V2_API_PATH);
+        String msg = generateContentJson(alertData);
+        HttpPost httpPost = constructHttpPost(v2Path, msg);
+        CloseableHttpClient httpClient = getDefaultClient();
```
Comment (Radeity) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: Use `try-with-resources` is better.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+                return resp;
+            }
+
+            log.error("send prometheus alert manager message error, return http status code: {} ", statusCode);
+            try {
+                HttpEntity entity = response.getEntity();
+                resp = EntityUtils.toString(entity, "utf-8");
+                EntityUtils.consume(entity);
+            } finally {
+                response.close();
```
Comment (Radeity) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: Ditto.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+            }
+            log.info("Prometheus alert manager send alert failed, title: {} ,content: {}, resp: {}",
```
Comment (Radeity) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: Duplicated error log, can we print log only once?

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        alertResult.setStatus("false");
+
+        if (Objects.equals(resp, AlertManagerConstants.ALERT_SUCCESS)) {
+            alertResult.setStatus("true");
+            alertResult.setMessage("prometheus alert manager send success");
+            return alertResult;
+        }
+
+        alertResult.setMessage("prometheus alert manager send fail");
+        log.info("send prometheus alert manager msg error,resp error");
```
Comment (Radeity) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: Ditto.

**Comment by Radeity** [comment]
In addition, file name in package `org.apache.dolphinscheduler.plugin.alert.prometheus` should start with `Prometheus`, like `PrometheusAlertChannel` and `PrometheusSender`.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-api/src/main/java/org/apache/dolphinscheduler/alert/api/AlertInputTips.java (hunk 3 above)
Comment (xjlgod) [reply dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-api/src/main/java/org/apache/dolphinscheduler/alert/api/AlertInputTips.java]: Ok.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java (hunk 5 above)
Comment (xjlgod) [reply dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: Ok.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java (hunk 4 above)
Comment (xjlgod) [reply dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: Ok.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java (hunk 7 above)
Comment (xjlgod) [reply dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: Ok.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java (hunk 8 above)
Comment (xjlgod) [reply dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/AlertManagerSender.java]: Ok.

**Comment by xjlgod** [comment]
> In addition, file name in package `org.apache.dolphinscheduler.plugin.alert.prometheus` should start with `Prometheus`, like `PrometheusAlertChannel` and `PrometheusSender`.


OK

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/PrometheusAlertSender.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    public AlertResult sendMessage(AlertData alertData) {
+        AlertResult alertResult;
+        try {
+            String resp = sendMsg(alertData);
+            return checkSendAlertManageMsgResult(resp);
+        } catch (Exception e) {
+            log.error("send prometheus alert manager alert msg  exception: {}", e.getMessage());
+            alertResult = new AlertResult();
+            alertResult.setStatus("false");
+            alertResult.setMessage("send prometheus alert manager alert fail.");
```
Comment (qingwli) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/PrometheusAlertSender.java]: Please add error into the message

Comment (xjlgod) [reply dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/PrometheusAlertSender.java]: CC

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/PrometheusAlertSender.java [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        annotations = config.get(PrometheusAlertConstants.NAME_ALERT_MANAGER_ANNOTATIONS);
+    }
+
+    public AlertResult sendMessage(AlertData alertData) {
+        AlertResult alertResult;
+        try {
+            String resp = sendMsg(alertData);
+            return checkSendAlertManageMsgResult(resp);
+        } catch (Exception e) {
+            String errorMsg = String.format("send prometheus alert manager alert msg  exception: %s", e.getMessage());
```
Comment (qingwli) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/PrometheusAlertSender.java]:
```suggestion
            String errorMsg = String.format("send prometheus alert manager alert error, exception: %s", e.getMessage());
```

**Review by qingwli — COMMENTED** [review]
Others LGTM

**Review by qingwli — APPROVED** [review]
LGTM

**Comment by qingwli** [comment]
PTAL @Radeity

**Review by Radeity — APPROVED** [review]

**Pull request merged**